flink-runtime/src/main/java/org/apache/flink/runtime/io/network/ShuffleEnvironment.java

```diff
@@ -0,0 +1,163 @@
+ * The task can request next available memory buffers from created here {@link ResultPartitionWriter}s to write shuffle data
+ * and buffers from created here {@link InputGate}s to read it.
+ *
+ * <h2>Service lifecycle management.</h2>
+ *
+ * <p>The interface contains method's to manage the lifecycle of the local shuffle environment:
+ * <ol>
+ *     <li>{@code start} is called when the {@link TaskExecutor} is being started.</li>
+ *     <li>{@code shutdown} is called when the {@link TaskExecutor} is being stopped.</li>
+ * </ol>
+ *
+ * <h2>Shuffle Input/Output management.</h2>
+ *
+ * <h3>Result partition management.</h3>
+ *
+ * <p>The interface implements a factory of result partition writers for the task output: {@code createResultPartitionWriters}.
+ * The created writers are grouped per task and handed over to the task thread upon its startup.
+ * The task is responsible for the writers lifecycle from that moment.
+ *
+ * <p>Partitions are released in the following cases:
+ * <ol>
+ *     <li>{@link ResultPartitionWriter#fail(Throwable)} and {@link ResultPartitionWriter#close()} are called
+ *     if the production has failed.</li>
+ *     <li>{@link ResultPartitionWriter#finish()} and {@link ResultPartitionWriter#close()} are called
+ *     if the production is done. The actual release can take some time
+ *     if 'the end of consumption' confirmation is being awaited implicitly
+ *     or the partition is later released by {@code releasePartitions(Collection<ResultPartitionID>)}.</li>
+ *     <li>{@code releasePartitions(Collection<ResultPartitionID>)} is called outside of the task thread,
+ *     e.g. to manage the local resource lifecycle of external partitions which outlive the task production.</li>
+ * </ol>
+ * The partitions, which currently still occupy local resources, can be queried with {@code updatePartitionInfo}.
+ *
+ * <h3>Input gate management.</h3>
+ *
+ * <p>The interface implements a factory for the task input gates: {@code createInputGates}.
+ * The created gates are grouped per task and handed over to the task thread upon its startup.
+ * The task is responsible for the gates lifecycle from that moment.
+ *
+ * <p>When tha task is deployed and the input gates are created, it can happen that not all consumed partitions
+ * are known at that moment e.g. because their producers have not been started yet.
+ * Therefore, the {@link ShuffleEnvironment} provides a method {@code updatePartitionInfo} to update them
+ * externally, ouside of the task thread, when the producer becomes known.
+ */
+public interface ShuffleEnvironment {
+
+	/**
+	 * Starts the internal related services upon {@link TaskExecutor}'s startup.
+	 *
+	 * @return a port to connect to the task executor for shuffle data exchange, -1 if only local connection is possible.
+	 */
+	int start() throws IOException;
+
+	/**
+	 * Shutdown the internal related services upon {@link TaskExecutor}'s stop.
+	 */
+	void shutdown();
+
+	boolean isShutdown();
```

Comment:
Also add a javadoc for this method to keep consistent with others?

Reply:
I removed it as discussed before